File: microservers/billetterieservice/src/main/java/io/multi/billetterieservice/domain/Offre.java
package io.multi.billetterieservice.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.OffsetDateTime;

/**
 * Entité représentant une offre de transport.
 * Une offre lie un trajet, un véhicule et un utilisateur propriétaire.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Offre {

    // ========== Champs de base ==========
    private Long offreId;
    private String offreUuid;
    private String tokenOffre;
    private Long trajetId;
    private Long vehiculeId;
    private Long userId;
    private LocalDate dateDepart;
    private LocalTime heureDepart;
    private LocalTime heureArriveeEstimee;
    private Integer nombrePlacesTotal;
    private Integer nombrePlacesDisponibles;
    private Integer nombrePlacesReservees;
    private BigDecimal montant;
    private BigDecimal montantPromotion;
    private String devise;
    private String statut;
    private Integer niveauRemplissage;
    private String pointRendezvous;
    private String conditions;
    private Boolean annulationAutorisee;
    private Integer delaiAnnulationHeures;
    private OffsetDateTime datePublication;
    private OffsetDateTime dateCloture;
    private OffsetDateTime dateDepartEffectif;
    private OffsetDateTime dateArriveeEffective;
    private OffsetDateTime createdAt;
    private OffsetDateTime updatedAt;

    // ========== Champs du Trajet (jointure) ==========
    private String trajetUuid;
    private String trajetLibelle;
    private BigDecimal trajetDistanceKm;
    private Integer trajetDureeMinutes;

    // ========== Champs du Départ (jointure via trajet) ==========
    private String departUuid;
    private String departLibelle;
    private String siteDepart;
    private String villeDepartLibelle;
    private String villeDepartUuid;
    private String regionDepartLibelle;

    // ========== Champs de l'Arrivée (jointure via trajet) ==========
    private String arriveeUuid;
    private String arriveeLibelle;
    private String siteArrivee;
    private String villeArriveeLibelle;
    private String villeArriveeUuid;
    private String regionArriveeLibelle;

    // ========== Champs du Véhicule (jointure) ==========
    private String vehiculeUuid;
    private String vehiculeImmatriculation;
    private String vehiculeMarque;
    private String vehiculeModele;
    private String vehiculeCouleur;
    private Integer vehiculeNombrePlaces;
    private Boolean vehiculeClimatise;
    private String vehiculeStatut;
    private String typeVehiculeLibelle;
    private String nomChauffeur;
    private String contactChauffeur;

    // ========== Champs de l'Utilisateur propriétaire (jointure) ==========
    private String userUuid;
    private String userUsername;
    private String userFullName;
    private String userEmail;
    private String userPhone;

    // ========== Champs calculés ==========
    /**
     * Retourne le montant effectif (promotion ou montant normal)
     */
    public BigDecimal getMontantEffectif() {
        if (montantPromotion != null && montantPromotion.compareTo(BigDecimal.ZERO) > 0) {
            return montantPromotion;
        }
        return montant;
    }

    /**
     * Vérifie si l'offre a une promotion active
     */
    public boolean hasPromotion() {
        return montantPromotion != null && montantPromotion.compareTo(BigDecimal.ZERO) > 0
                && montantPromotion.compareTo(montant) < 0;
    }

    /**
     * Calcule le pourcentage de réduction
     */
    public BigDecimal getPourcentageReduction() {
        if (!hasPromotion()) {
            return BigDecimal.ZERO;
        }
        return montant.subtract(montantPromotion)
                .multiply(new BigDecimal("100"))
                .divide(montant, 2, java.math.RoundingMode.HALF_UP);
    }
}

File: mobile_app/lib/models/offre.dart
class Offre {
  final String offreUuid;
  final String? tokenOffre;
  final String? trajetLibelle;
  final int? trajetDureeMinutes;
  final String? departLibelle;
  final String? siteDepart;
  final String? villeDepartLibelle;
  final String? villeDepartUuid;
  final String? arriveeLibelle;
  final String? siteArrivee;
  final String? villeArriveeLibelle;
  final String? villeArriveeUuid;
  final String? dateDepart;
  final String? heureDepart;
  final String? heureArriveeEstimee;
  final int nombrePlacesDisponibles;
  final int nombrePlacesTotal;
  final double montant;
  final double? montantPromotion;
  final String? devise;
  final String? vehiculeImmatriculation;
  final String? vehiculeMarque;
  final String? vehiculeModele;
  final bool vehiculeClimatise;
  final String? typeVehiculeLibelle;
  final String? nomChauffeur;
  final String? contactChauffeur;
  final double? noteMoyenne;
  final int? nombreAvis;
  final String? statut;
  final String? pointRendezvous;
  final String? conditions;
  final bool annulationAutorisee;
  final int? delaiAnnulationHeures;

  Offre({
    required this.offreUuid,
    this.tokenOffre,
    this.trajetLibelle,
    this.trajetDureeMinutes,
    this.departLibelle,
    this.siteDepart,
    this.villeDepartLibelle,
    this.villeDepartUuid,
    this.arriveeLibelle,
    this.siteArrivee,
    this.villeArriveeLibelle,
    this.villeArriveeUuid,
    this.dateDepart,
    this.heureDepart,
    this.heureArriveeEstimee,
    this.nombrePlacesDisponibles = 0,
    this.nombrePlacesTotal = 0,
    this.montant = 0,
    this.montantPromotion,
    this.devise,
    this.vehiculeImmatriculation,
    this.vehiculeMarque,
    this.vehiculeModele,
    this.vehiculeClimatise = false,
    this.typeVehiculeLibelle,
    this.nomChauffeur,
    this.contactChauffeur,
    this.noteMoyenne,
    this.nombreAvis,
    this.statut,
    this.pointRendezvous,
    this.conditions,
    this.annulationAutorisee = true,
    this.delaiAnnulationHeures,
  });

  double get montantEffectif =>
      (montantPromotion != null && montantPromotion! > 0)
          ? montantPromotion!
          : montant;

  bool get hasPromotion =>
      montantPromotion != null &&
      montantPromotion! > 0 &&
      montantPromotion! < montant;

  String get vehiculeDescription {
    if (vehiculeMarque != null && vehiculeModele != null) {
      return '$vehiculeMarque $vehiculeModele';
    }
    return typeVehiculeLibelle ?? 'Véhicule';
  }

  String get durationFormatted {
    if (trajetDureeMinutes == null) return '';
    final h = trajetDureeMinutes! ~/ 60;
    final m = trajetDureeMinutes! % 60;
    if (h > 0 && m > 0) return '${h}h${m.toString().padLeft(2, '0')}';
    if (h > 0) return '${h}h';
    return '${m}min';
  }

  String get heuresDepartFormatted =>
      heureDepart?.substring(0, 5) ?? '';

  String get heuresArriveeFormatted =>
      heureArriveeEstimee?.substring(0, 5) ?? '';

  factory Offre.fromJson(Map<String, dynamic> json) {
    return Offre(
      offreUuid: json['offreUuid'] ?? '',
      tokenOffre: json['tokenOffre'],
      trajetLibelle: json['trajetLibelle'],
      trajetDureeMinutes: json['trajetDureeMinutes'],
      departLibelle: json['departLibelle'],
      siteDepart: json['siteDepart'],
      villeDepartLibelle: json['villeDepartLibelle'],
      villeDepartUuid: json['villeDepartUuid'],
      arriveeLibelle: json['arriveeLibelle'],
      siteArrivee: json['siteArrivee'],
      villeArriveeLibelle: json['villeArriveeLibelle'],
      villeArriveeUuid: json['villeArriveeUuid'],
      dateDepart: json['dateDepart'],
      heureDepart: json['heureDepart'],
      heureArriveeEstimee: json['heureArriveeEstimee'],
      nombrePlacesDisponibles: json['nombrePlacesDisponibles'] ?? 0,
      nombrePlacesTotal: json['nombrePlacesTotal'] ?? 0,
      montant: (json['montant'] ?? 0).toDouble(),
      montantPromotion: json['montantPromotion']?.toDouble(),
      devise: json['devise'],
      vehiculeImmatriculation: json['vehiculeImmatriculation'],
      vehiculeMarque: json['vehiculeMarque'],
      vehiculeModele: json['vehiculeModele'],
      vehiculeClimatise: json['vehiculeClimatise'] ?? false,
      typeVehiculeLibelle: json['typeVehiculeLibelle'],
      nomChauffeur: json['nomChauffeur'],
      contactChauffeur: json['contactChauffeur'],
      noteMoyenne: json['noteMoyenne']?.toDouble(),
      nombreAvis: json['nombreAvis'],
      statut: json['statut'],
      pointRendezvous: json['pointRendezvous'],
      conditions: json['conditions'],
      annulationAutorisee: json['annulationAutorisee'] ?? true,
      delaiAnnulationHeures: json['delaiAnnulationHeures'],
    );
  }
}
